Reference lc

rbiswas4 opened the pull request:
Added reference light curves:

Added a class which calculates truth parameters for a model and when supplied with a set of observation al parameters and hardware/site parameters, and ID values corresponding to SN in an instance catalog, can pull the model parameters describing the astrophysical object from the database and compute the light curve. 

The source code to do this is in monitor/truth.py . An example notebook demonstrating the usage is in example/reference_lc.ipynb

tests need to be written before we can merge this.

It would be good if people, pariticularly @jchiang87 and @jbkalmbach can look at the API and example and let me know if there are suggestions/requirements for improving the API for their work. Once the API is agreed upon, I would like to move most of this to an abstract class and subclass different objects from there.

rbiswas4 committed:
Added file truths.py to access SN parameters from fatboy.
-  First basic  function provided checked that it works with the script
	new file:   truths.py

rbiswas4 committed:
added code to get SNOBject

rbiswas4 committed:
Working code
	modified:   truths.py

rbiswas4 committed:
new file:   reference_lc.ipynb

rbiswas4 committed:
Modified Added reference light Curves
	modified:   reference_lc.ipynb

rbiswas4 committed:
Added dataset for Twinkles
	new file:   SelectedKrakenVisits.csv

rbiswas4 committed:
Added some functionality and added docstrings

- Added Functionality: obtain all ids in the table, find the number of
  objects
- Added docstrings and removed script part
	modified:   __init__.py
	new file:   truth.py
	deleted:    truths.py

rbiswas4 committed:
Fixed bug pointed out by @jchiang
- Added method to go from catsim uniqueID to table ID
	modified:   truth.py

rbiswas4 committed:
Fixed bug pointed out by @jchiang
	modified:   truth.py
The idSequence or idValue was pointing to the ID in the table for the
astro_obj Table for example TwinkSN. We would like the IDs to point to
the unique ID in catsim, across all astrophysical objects. The two are
connected by a bit shift and addition, so this problem is taken care of
by inverting this operation through a static function
```
def uniqueIDtoTableId
```
Later on, we will change this to directly use the functionality from
catsim

rbiswas4 committed:
Fixed demo example file
	modified:   examples/reference_lc.ipynb

rbiswas4 committed:
Merge branch 'master' into reference_lc
Fixed conflict

rbiswas4 committed:
Modified the sql query built so that all parameters in the table can be
obtained. (There was a bug involving the conversion of Uinique IR
refIDCol  to IdCol when the idSequence passed was None)
	modified:   truth.py

jchiang87 commented:
Hi @rbiswas4,  Thanks for this work!  I'll comment on the API after I've had a chance to play with the code.   I did note that in your notebook example, the line
```
data_dir = os.path.join(monitor.__file__.split('/python')[0], 'data')
```
can be replaced by
```
data_dir = os.path.join(os.environ['MONITOR_DIR'], 'data')
```
since eups defines that env var as part of the set up.

jchiang87 commented:
So I just did this
```
[edison01] pwd
/global/homes/j/jchiang8/twinkles
[edison01] chmod -R go+rx inputs/
[edison01] chmod -R go+rx Run1.1/
```
Is that sufficient?  (It looked to me as though you should have already had access.)

rbiswas4 commented:
Yes, I noticed later that this is how you had written tests,  and I had forgotten about eups. I should change the way I had the code.

I think it would be good if we can provide a data directory variable via the __init__ or setup file so that one can just always do:
```
os.path.join(monitor.data, datafile)
```

Incidentally,  within LSST we have been moving to 
```
from lsst.utils import getPackageDir
getPackageDir(eups_pkgDir) 
``` 
to get eups tailored exceptions instead of using `os.environ`

jchiang87 commented:
Regarding the interface, I'd like to see the db connection and opsim csv stuff encapsulated, e.g., for my light curve overlay work, I immediately wrote this wrapper class:
```
class RefLightCurveServer(object):
    def __init__(self, idSequence, opsim_csv=None, db_config=None):
        if opsim_csv is None:
            opsim_csv = os.path.join(os.environ['MONITOR_DIR'], 'data',
                                     'SelectedKrakenVisits.csv')
        df = pd.read_csv(opsim_csv, index_col='obsHistID')
        self.opsim_df = df[['expMJD', 'filter', 'fiveSigmaDepth']]

        if db_config is None:
            db_config = os.path.join(lsst.utils.getPackageDir('sims_catUtils'),
                                     'config', 'db.py')
        config = bcm.BaseCatalogConfig()
        config.load(db_config)
        self._create_connection(config)
        self.reflc = RefLightCurves(idSequence=idSequence,
                                    tableName='TwinkSN',
                                    dbConnection=self.connection,
                                    dbCursor=self.cursor)

        self.lsstBP = BandpassDict.loadBandpassesFromFiles()[0]

    def _create_connection(self, config):
        username = dafPersist.DbAuth.username(config.host, config.port)
        password = dafPersist.DbAuth.password(config.host, config.port)
        self.connection = pymssql.connect(user=username,
                                          password=password,
                                          database=config.database,
                                          port=config.port,
                                          host=config.host)
        self.cursor = self.connection.cursor()

    def lightCurve(self, idValue, band):
        lc = self.reflc.lightCurve(idValue, self.opsim_df, self.lsstBP)
        my_band = lc[lc['band'] == band]
        return my_band[['time', 'flux', 'fluxerr']]
```
Something like the above `__init__` code could be included in your new base class.   As far as serving up the light curves themselves are concerned, providing them as `desc.monitor.LightCurve` objects would be fine, but I'd like to expand the interface of the `LightCurve` class to make it easier to get the `time`, `flux`, `fluxerr` arrays for a specified band a little easier (e.g., as in my `lightCurve` method above).  It also might be desirable to redesign `LightCurve` so that the `build*` methods don't appear in contexts where they wouldn't be used (i.e., for these reference light curves).  We could do this via a class hierarchy or by stripping down the `LightCurve` class and use factory methods to create and fill the instances with data.  This latter option would jibe with my hope to make those objects immutable.

drphilmarshall commented:
@jchiang87 Your `RefLightCurveServer` object seems to do a subset of what the `SNMonitor` object is supposed to do, I think. Imagine:
```
We = SNMonitor()
We.read_in_opsim_data()
We.create_catsim_db_connection()
We.get_observed_light_curve(ID)
We.get_reference_light_curve(ID)
We.plot_light_curves()
We.compare_light_curves()
```
What do you think?

jchiang87 commented:
I think that's fine, but I guess if the `.read_in_opsim_data(), .create_catsim_db_connection()` calls are boilerplate, I would just stick them in the SNMonitor constructor and not call them separately.  Are the `.get_*` functions meant to fetch data not already loaded?  Maybe they could likewise be encapsulated in the `.plot_light_curves()` function and the calling sequence would then be something like
```
We = SNMonitor()
We.plot_light_curves(obj=objectId, ref=idValue)
```
where the specific light curves are plotted (and overlayed) if the keywords are given.  What would `.compare_light_curves()` do aside from plotting overlays? Plotting residuals/ratios?

jchiang87 commented on python/desc/monitor/truth.py:
```diff
@@ -0,0 +1,305 @@
+        Returns
+        -------
+        astropy.Table containing the following columns at the minimum
+        ['time', 'flux', 'fluxerr', 'band']. If a set of indexes are provided
```
It would be good to state here that the flux is in maggies and the time is in MJD.

rbiswas4 replied:
Yes! absolutely.

drphilmarshall commented:
Agreed on the boilerplate stuff - `__init__()` methods take kwargs after all. 
BTW in my obsessively literate python style I'd want  to write
```python
We.plot_light_curves(for=objectId, comparing_with=idValue)
```
:-)  Ideally there would only be one ID though: I don't want to have to know in advance which reference object to choose! 
```python
We.plot_light_curves(for=objectId, comparing_with='reference')
```
should Do The Right Thing, for example...

rbiswas4 commented:
@jchiang87 

Thanks for your comments. We could definitely add functionality to get those things in (I did not follow the broader aim you have about making light curve objects immutable, and not have build methods and would be happy to hear more about it, but adding per band light curves is a concrete thing we can add right away). 

There is a somewhat broader question that would be good to discuss more generally: Should we aim to make these tools entirely LSST specific or should they be broader?  At some level, it does not make sense to have tools that are Twinkles specific to do anything more than LSST. Then again, it might be useful to play with.

rbiswas4 commented:
@drphilmarshall 

If ObjectID is a source/forcedSource ID, then there is no guarantee that there will be a unique matched IdValue is a uniqueID in the simulation with zero uncertainty.  In fact, if you think about false positives, there will be ObjectID s that have no corresponding (true) uniqueID.  Are we thinking about the same thing?

jchiang87 commented:
That's pretty obsessive, @drphilmarshall !  Since the `for` and `comparing_with` keywords would be used in the underlying code, those names might make the internal implementation somewhat less literate (especially since `for` is reserved, looping would be a challenge).

drphilmarshall commented:
@jchiang I know, that is the price you pay for sparkling prose at the top
level. It pains me.

@rbiswas4 I guess I was imagining that the SNMonitor would take care of the
matching of DM sources and input objects, somehow. In your current model, I
have to employ some other system to tell me which ID to compare with each
DM source.

On Fri, Jun 17, 2016 at 3:44 PM, James Chiang <notifications@github.com>
wrote:

> That's pretty obsessive, @drphilmarshall
> <https://github.com/drphilmarshall> ! Since the for and comparing_with
> keywords would be used in the underlying code, those names might make the
> internal implementation somewhat less literate (especially since for is
> reserved, looping would be a challenge).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DarkEnergyScienceCollaboration/Monitor/pull/31#issuecomment-226898683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AArY9xpZzZ049uHxfOZSjySIJgZNaj5Bks5qMyNfgaJpZM4I4AZf>
> .
>

rbiswas4 committed:
Added connection features as requested by @jchiang
	modified:   python/desc/monitor/truth.py

rbiswas4 committed:
Chnaged the truth.py module to incorporate requests in response to the
PR: The main changes are:

- The original design of refLightCurve is slightly changed to include
  variables like hostname, observations, and bandpasses which can all
  be passed in with default values to retrieve the older behavior.
- Addition of a class method to enable quick and simple analysis of
  Twinkles data by appropriately using defaults for the Twinkles data
  for a number of parameters
- Addition of a paameter `bandName` in the method
  `monitor.RefLightCurves.lightCurve(idValue, bandName=None, ...)`.
  When supplied a parameter bandName='r', this method only returns
  the light curve in the single specified band.
- Added a parameter dbHostName in both the original design and the
  class method to change the default dbHostName from localhost
  (specified by default and appropriate for work with ssh tunnels)
  to a user specified value (useful for use with whitelisted machines)
- Updated and added docstrings to the class, Atrributes and methods.
- Demonstration of these methods in `examples/reference_lc.ipynb`.

	modified:   python/desc/monitor/truth.py
	modified:   examples/reference_lc.ipynb

jchiang87 commented:
This looks good to go.  Let's merge it.

drphilmarshall commented:
Excellent! Nice work, @rbiswas4 :-)

On Mon, Jun 20, 2016 at 9:43 AM, James Chiang <notifications@github.com>
wrote:

> This looks good to go. Let's merge it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DarkEnergyScienceCollaboration/Monitor/pull/31#issuecomment-227198434>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AArY91TcSMnaDWr__vvWh9PpJObhBCOQks5qNsMUgaJpZM4I4AZf>
> .
>

rbiswas4 commented:
@drphilmarshall 
>  I guess I was imagining that the SNMonitor would take care of the
matching of DM sources and input objects, somehow. In your current model, I
have to employ some other system to tell me which ID to compare with each
DM source.

I can see that we would want to build such a functionality into Monitor, and we will do something like that eventually. I am just saying that we need perhaps more optional arguments, since we have to account for the possibility that after attempting to do the right thing, the code will fail in certain situations.

jchiang87 commented:
In my example, I was also considering the likely use case where one might want to plot the reference light curve by itself, e.g., before the level 2 pipeline has been run.
